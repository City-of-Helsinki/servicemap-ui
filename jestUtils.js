import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { ThemeProvider } from '@mui/material/styles';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import themes from './src/themes';
import english from './src/i18n/en';

const mockStore = configureStore([]);

// Mock props for intl provider
const intlMock = {
  locale: 'en',
  messages: english,
  wrapRichTextChunksInFragment: false,
};

const Providers = (mockState) => ({ children }) => {
  const store = mockStore(mockState);
  return (
    <Provider store={store}>
      <IntlProvider {...intlMock}>
        <ThemeProvider theme={themes.SMTheme}>
          {children}
        </ThemeProvider>
      </IntlProvider>
    </Provider>
  );
}

export const getRenderWithProviders = mockState => component => render(component, { wrapper: Providers(mockState) });
