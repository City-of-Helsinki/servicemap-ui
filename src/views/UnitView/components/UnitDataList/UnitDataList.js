import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import {
  EventItem,
  ReservationItem,
  ServiceItem,
  TitledList,
} from '../../../../components';

const UnitDataList = ({
  data, listLength, type, period, disableTitle, navigator,
}) => {
  const location = useLocation();
  const unit = useSelector(state => state.selectedUnit.unit.data);

  const dataItems = data.data;
  let fullDataLength;

  if (type === 'educationServices') {
    fullDataLength = dataItems?.length;
  } else {
    fullDataLength = data.max;
  }

  const { isFetching } = data;

  if (!dataItems) {
    return null;
  }
  const endIndex = listLength > dataItems.length ? dataItems.length : listLength;
  const shownData = dataItems.length ? dataItems.slice(0, endIndex) : null;
  const showButton = fullDataLength > listLength;

  const onButtonClick = () => {
    if (navigator) {
      navigator.replace({
        ...location,
        hash: `Unit${type}Button`,
      });
      navigator.push('unit', { id: unit.id, type, period });
    }
  };

  const renderListItems = () => {
    if (type === 'events') {
      return (
        shownData.map((item, i) => (
          <EventItem simpleItem key={item.id} event={item} divider={i !== shownData.length - 1} />
        ))
      );
    } if (type === 'reservations') {
      return (
        shownData.map((item, i) => (
          <ReservationItem key={item.id} reservation={item} divider={i !== shownData.length - 1} />
        ))
      );
    }
    if (type === 'services' || type === 'educationServices') {
      return (
        shownData.map((item, i) => (
          <ServiceItem
            key={item.id}
            service={item}
            divider={i !== shownData.length - 1}
          />
        ))
      );
    }
    return null;
  };

  if (unit && shownData && shownData.length) {
    return (
      <div>
        <TitledList
          title={!disableTitle ? <FormattedMessage id={`unit.${type}`} /> : null}
          description={<FormattedMessage id={`unit.${type}.description`} values={period ? { period } : null} />}
          detailedTitle
          divider={false}
          titleComponent="h4"
          buttonMessageID={`unit.${type}.more`}
          buttonMessageCount={fullDataLength - listLength}
          loading={isFetching}
          buttonID={`Unit${type}Button`}
          onButtonClick={showButton ? onButtonClick : null}
        >
          {renderListItems()}
        </TitledList>
      </div>
    );
  } return (
    null
  );
};

UnitDataList.propTypes = {
  data: PropTypes.objectOf(PropTypes.any),
  listLength: PropTypes.number,
  type: PropTypes.string.isRequired,
  navigator: PropTypes.objectOf(PropTypes.any),
  period: PropTypes.string,
  disableTitle: PropTypes.bool,
};

UnitDataList.defaultProps = {
  data: null,
  listLength: 5,
  navigator: null,
  period: null,
  disableTitle: false,
};

export default UnitDataList;
