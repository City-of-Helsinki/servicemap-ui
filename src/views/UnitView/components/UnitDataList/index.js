import { withStyles } from '@mui/styles';
import styles from '../../styles/styles';
import UnitDataList from './UnitDataList';

export default withStyles(styles)(UnitDataList);
