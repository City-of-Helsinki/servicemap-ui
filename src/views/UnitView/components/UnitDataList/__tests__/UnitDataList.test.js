import React from 'react';
import { getRenderWithProviders } from '../../../../../../jestUtils';
import UnitDataList from '../UnitDataList';


const mockData = { // Data for testing unit services
  max: 5,
  data: [
    {
      id: 'helmet:241163',
      name: { fi: 'Suomen kielen kielikahvila' },
    },
    {
      id: 'helmet:241975',
      name: { fi: 'R&A x Armas: Vanhenemisen estetiikka' },
    },
    {
      id: 'helmet:239092',
      name: {
        fi: 'Tietotekniikkaopastusta senioreille',
        sv: 'Digital handledning',
        en: 'Digital guidance for seniors',
      },
    },
    {
      id: 'helmet:242249',
      name: { fi: 'Marttojen neulontaklinikka: kaarrokeneule omille mitoille' },
    },
    {
      id: 'helmet:242089',
      name: { fi: 'Marttojen neulontaklinikka' },
    },
  ],
};

const mockData2 = { // Data for testing educational unit services
  data: [
    {
      clarification: null,
      name: {
        fi: 'A-ranska (lukion kieliohjelma)',
        sv: 'A-franska (gymnasiets språkprogram)',
        en: 'A-French (upper secondary school language programme)',
      },
      root_service_node: 1087,
      period: null,
      id: 10,
    },
    {
      clarification: null,
      name: {
        fi: 'A-ruotsi (lukion kieliohjelma)',
        sv: 'A-svenska (gymnasiets språkprogram)',
        en: 'A-Swedish (upper secondary school language programme)',
      },
      root_service_node: 1087,
      period: null,
      id: 11,
    },
    {
      clarification: null,
      name: {
        fi: 'A-saksa (lukion kieliohjelma)',
        sv: 'A-tyska (gymnasiets språkprogram)',
        en: 'A-German (upper secondary school language programme)',
      },
      root_service_node: 1087,
      period: null,
      id: 12,
    },
    {
      clarification: null,
      name: {
        fi: 'A-suomi (lukion kieliohjelma)',
        sv: 'A-finska (gymnasiets språkprogram)',
        en: 'A-Finnish (upper secondary school language programme)',
      },
      root_service_node: 1087,
      period: null,
      id: 13,
    },
    {
      clarification: null,
      name: {
        fi: 'B-englanti (lukion kieliohjelma)',
        sv: 'B-engelska (gymnasiets språkprogram)',
        en: 'B-English (upper secondary school language programme)',
      },
      root_service_node: 1087,
      period: null,
      id: 88,
    },
    {
      clarification: null,
      name: {
        fi: 'B-espanja (lukion kieliohjelma)',
        sv: 'B-spanska (gymnasiets språkprogram)',
        en: 'B-Spanish (upper secondary school language programme)',
      },
      root_service_node: 1087,
      period: null,
      id: 89,
    },
  ],
};


const mockProps = { // Props for testing unit services
  data: mockData,
  listLength: 3,
  type: 'events',
};

const mockProps2 = { // Props for testing educational unit services
  data: mockData2,
  listLength: 3,
  type: 'educationServices',
};

const renderWithProviders = getRenderWithProviders({
  selectedUnit: { unit: { data: { id: 51342 } } },
  user: { locale: 'en' },
  service: { current: null },
});


describe('<UnitDataList />', () => {
  it('should work', () => {
    const { container } = renderWithProviders(<UnitDataList {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('does render events list with correct items', () => {
    const { getAllByTestId } = renderWithProviders(<UnitDataList {...mockProps} />);
    const count = getAllByTestId('EventItem').length;
    expect(count === mockProps.listLength).toBeTruthy();
  });

  it('renders correct event count number in button', () => {
    const { getByText } = renderWithProviders(<UnitDataList {...mockProps} />);
    const moreButton = getByText('Show more events', { exact: false });
    const numberInButtonText = parseInt(moreButton.textContent.match(/\d+/)[0], 10);
    const { data, listLength } = mockProps;
    expect(numberInButtonText === (data.max - listLength)).toBeTruthy();
  });

  it('renders correct services count number in schoool services button', () => {
    const { getByText } = renderWithProviders(<UnitDataList {...mockProps2} />);
    const moreButton = getByText('Show more services', { exact: false });
    const numberInButtonText = parseInt(moreButton.textContent.match(/\d+/)[0], 10);
    const { data, listLength } = mockProps2;
    expect(numberInButtonText === (data.data.length - listLength)).toBeTruthy();
  });
});
