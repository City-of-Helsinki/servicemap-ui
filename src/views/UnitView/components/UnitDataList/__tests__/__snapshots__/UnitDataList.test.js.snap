// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UnitDataList /> should work 1`] = `
<div>
  <div>
    <div
      class="TitledList-titleContainer-6 TitledList-marginHorizontal-4"
    >
      <h4
        class="MuiTypography-root MuiTypography-subtitle1 TitledList-left-1 TitledList-marginVertical-3 css-dlbwt1-MuiTypography-root"
      >
        Events
      </h4>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-alignLeft TitledList-description-5 css-10dderv-MuiTypography-root"
    >
      Here you can find information about events provided by the unit
    </p>
    <ul
      class="MuiList-root css-1mk9mw3-MuiList-root"
    >
      <li
        class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button ResultItem-listItem-19 css-16e6fe9-MuiButtonBase-root-MuiListItem-root"
        data-testid="EventItem"
        role="link"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root ResultItem-listItemIcon-20 css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-tp7oeh-MuiSvgIcon-root"
            data-testid="EventIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
            />
          </svg>
        </div>
        <div
          class="ResultItem-itemTextContainer-26  "
        >
          <div
            class="ResultItem-topRow-28"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-24 ResultItem-srOnly css-1ak20dk-MuiTypography-root"
              style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
            >
              
    Suomen kielen kielikahvila 
     
     
     
    
  
            </p>
            <p
              aria-hidden="true"
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-24    ResultItem-title css-1ak20dk-MuiTypography-root"
              role="textbox"
            >
              Suomen kielen kielikahvila
            </p>
          </div>
        </div>
      </li>
      <li
        aria-hidden="true"
      >
        <hr
          class="MuiDivider-root MuiDivider-inset ResultItem-divider-34 css-uls1yk-MuiDivider-root"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button ResultItem-listItem-19 css-16e6fe9-MuiButtonBase-root-MuiListItem-root"
        data-testid="EventItem"
        role="link"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root ResultItem-listItemIcon-20 css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-tp7oeh-MuiSvgIcon-root"
            data-testid="EventIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
            />
          </svg>
        </div>
        <div
          class="ResultItem-itemTextContainer-26  "
        >
          <div
            class="ResultItem-topRow-28"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-24 ResultItem-srOnly css-1ak20dk-MuiTypography-root"
              style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
            >
              
    R&A x Armas: Vanhenemisen estetiikka 
     
     
     
    
  
            </p>
            <p
              aria-hidden="true"
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-24    ResultItem-title css-1ak20dk-MuiTypography-root"
              role="textbox"
            >
              R&A x Armas: Vanhenemisen estetiikka
            </p>
          </div>
        </div>
      </li>
      <li
        aria-hidden="true"
      >
        <hr
          class="MuiDivider-root MuiDivider-inset ResultItem-divider-34 css-uls1yk-MuiDivider-root"
        />
      </li>
      <li
        class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button ResultItem-listItem-19 css-16e6fe9-MuiButtonBase-root-MuiListItem-root"
        data-testid="EventItem"
        role="link"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root ResultItem-listItemIcon-20 css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-tp7oeh-MuiSvgIcon-root"
            data-testid="EventIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
            />
          </svg>
        </div>
        <div
          class="ResultItem-itemTextContainer-26  "
        >
          <div
            class="ResultItem-topRow-28"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-24 ResultItem-srOnly css-1ak20dk-MuiTypography-root"
              style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
            >
              
    Digital guidance for seniors 
     
     
     
    
  
            </p>
            <p
              aria-hidden="true"
              class="MuiTypography-root MuiTypography-body2 ResultItem-title-24    ResultItem-title css-1ak20dk-MuiTypography-root"
              role="textbox"
            >
              Digital guidance for seniors
            </p>
          </div>
        </div>
      </li>
    </ul>
    <button
      aria-label="Show more events (2)"
      class="MuiButtonBase-root SMButton SMButton-button-10 SMButton-smallButton-11 SMButton-margin-12  SMButton-default-16 css-fvdnvy-MuiButtonBase-root"
      id="UniteventsButton"
      role="link"
      tabindex="0"
      type="button"
      variant="contained"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium TitledList-buttonIcon-9 css-6yqi1c-MuiSvgIcon-root"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m7 10 5 5 5-5z"
        />
      </svg>
      <p
        aria-hidden="true"
        class="MuiTypography-root MuiTypography-button SMButton-typography-17 css-kkeniz-MuiTypography-root"
      >
        Show more events (2)
      </p>
    </button>
  </div>
</div>
`;
